src/main/java/org/sonatype/goodies/packageurl/PackageUrl.java

```diff
@@ -67,28 +64,29 @@
   private final String version;
 
   @Nullable
-  private final Map<String, String> qualifiers;
+  // sorted map of qualifiers (lexicographically); see: https://github.com/package-url/purl-spec/issues/51
+  private final SortedMap<String, String> qualifiers;
 
   @Nullable
   private final List<String> subpath;
 
   /**
    * Values should have already been validated via {@link PackageUrlBuilder} and {@link PackageUrlValidator}.
    */
-  @VisibleForTesting
+  // @VisibleForTesting
   PackageUrl(final String type,
              @Nullable final List<String> namespace,
              final String name,
              @Nullable final String version,
              @Nullable final Map<String, String> qualifiers,
              @Nullable final List<String> subpath)
   {
-    this.type = checkNotNull(type);
-    this.namespace = namespace != null ? ImmutableList.copyOf(namespace) : null;
-    this.name = checkNotNull(name);
+    this.type = requireNonNull(type);
+    this.namespace = namespace != null ? unmodifiableList(new ArrayList<>(namespace)) : null;
+    this.name = requireNonNull(name);
     this.version = version;
-    this.qualifiers = qualifiers != null ? ImmutableMap.copyOf(qualifiers) : null;
-    this.subpath = subpath != null ? ImmutableList.copyOf(subpath) : null;
+    this.qualifiers = qualifiers != null ? unmodifiableSortedMap(new TreeMap<>(qualifiers)) : null;
+    this.subpath = subpath != null ? unmodifiableList(new ArrayList<>(subpath)) : null;
```

Comment:
the ctor here is generally internal use only, so creating new collection copies here may be a waste, as these should already be protected from changes from the builder or parser; most uses of builder don't expect to mutate post build either, so I think the collection bits here may be wasteful?

Reply:
Done.